http://sourceware.org/ml/gdb-cvs/2010-08/msg00090.html

### src/gdb/testsuite/ChangeLog	2010/08/17 20:59:03	1.2412
### src/gdb/testsuite/ChangeLog	2010/08/17 21:31:12	1.2413
## -1,4 +1,10 @@
 2010-08-17  Jan Kratochvil  <jan.kratochvil@redhat.com>
+
+	* gdb.cp/infcall-dlopen.exp: New file.
+	* gdb.cp/infcall-dlopen.cc: New file.
+	* gdb.cp/infcall-dlopen-lib.cc: New file.
+
+2010-08-17  Jan Kratochvil  <jan.kratochvil@redhat.com>
 	    Pedro Alves  <pedro@codesourcery.com>
 
 	PR breakpoints/11371
--- src/gdb/testsuite/gdb.cp/infcall-dlopen-lib.cc
+++ src/gdb/testsuite/gdb.cp/infcall-dlopen-lib.cc	2011-01-01 12:52:07.599598000 +0000
@@ -0,0 +1,16 @@
+/* This testcase is part of GDB, the GNU debugger.
+
+   Copyright 2010 Free Software Foundation, Inc.
+
+   This program is free software; you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation; either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.  */
--- src/gdb/testsuite/gdb.cp/infcall-dlopen.cc
+++ src/gdb/testsuite/gdb.cp/infcall-dlopen.cc	2011-01-01 12:52:07.883051000 +0000
@@ -0,0 +1,37 @@
+/* This testcase is part of GDB, the GNU debugger.
+
+   Copyright 2010 Free Software Foundation, Inc.
+
+   This program is free software; you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation; either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.  */
+
+#include <dlfcn.h>
+#include <stddef.h>
+
+static int
+openlib (const char *filename)
+{
+  void *h = dlopen (filename, RTLD_LAZY);
+
+  if (h == NULL)
+    return 0;
+  if (dlclose (h) != 0)
+    return 0;
+  return 1;
+}
+
+int
+main (void)
+{
+  return 0;
+}
--- src/gdb/testsuite/gdb.cp/infcall-dlopen.exp
+++ src/gdb/testsuite/gdb.cp/infcall-dlopen.exp	2011-01-01 12:52:08.166496000 +0000
@@ -0,0 +1,46 @@
+# Copyright 2010 Free Software Foundation, Inc.
+
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+if {[skip_shlib_tests]} {
+    return 0
+}
+
+if [get_compiler_info not-used] {
+    return -1
+}
+
+set testfile "infcall-dlopen"
+set srcmainfile ${testfile}.cc
+set srclibfile ${testfile}-lib.cc
+set executable ${testfile}
+set libfile ${objdir}/${subdir}/${executable}.so
+set binfile ${objdir}/${subdir}/${executable}
+
+# Use completely arbitrary file for $libfile source.
+if { [gdb_compile_shlib ${srcdir}/${subdir}/${srclibfile} ${libfile} {debug c++}] != ""
+     || [prepare_for_testing ${testfile}.exp ${executable} ${srcmainfile} {debug c++ shlib_load}] } {
+    return -1
+}
+
+if { ![runto_main] } {
+    return -1
+}
+
+for {set i 0} {$i < 10} {incr i} {
+    gdb_test "p openlib (\"${libfile}\")" " = 1" "test $i"
+    # Try to exploit the GDB trashed memory.
+    gdb_test "b openlib" {Breakpoint [0-9]+ at .*} "test $i stub 1"
+    gdb_test_no_output {delete $bpnum} "test $i stub 2"
+}
